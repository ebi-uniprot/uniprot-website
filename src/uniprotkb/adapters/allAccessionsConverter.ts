import { AlternativeProductsComment, CommentType } from '../types/commentTypes';
import { UniProtkbAPIModel } from './uniProtkbConverter';

const convertAllAccessions = (data: UniProtkbAPIModel) => {
  const alternativeProducts = data.comments?.find(
    (comment) => comment.commentType === CommentType.ALTERNATIVE_PRODUCTS
  ) as AlternativeProductsComment;
  const isoforms = alternativeProducts.isoforms
    .map((isoform) => isoform.isoformIds.map((isoformId) => isoformId))
    .flat();
  return { isoforms };
};

export default convertAllAccessions;
