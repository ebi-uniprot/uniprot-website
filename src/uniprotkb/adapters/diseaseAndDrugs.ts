import { CommentType } from '../types/commentTypes';
import KeywordCategory from '../types/keywordCategory';
import FeatureType from '../types/featureType';
import { convertSection } from './sectionConverter';
import EntrySection from '../types/entrySection';
import { UniProtkbAPIModel } from './uniProtkbConverter';
import { Xref } from '../../shared/types/apiModel';

const keywordsCategories: KeywordCategory[] = ['Disease'];

const featuresCategories: FeatureType[] = ['Mutagenesis'];

const commentsCategories: CommentType[] = [
  'DISEASE',
  'ALLERGEN',
  'DISRUPTION PHENOTYPE',
  'TOXIC DOSE',
  'PHARMACEUTICAL',
];

const convertDiseaseAndDrugs = (
  data: UniProtkbAPIModel,
  uniProtKBCrossReferences?: Xref[]
) =>
  convertSection(
    data,
    commentsCategories,
    keywordsCategories,
    featuresCategories,
    EntrySection.DiseaseAndDrugs,
    uniProtKBCrossReferences
  );

export default convertDiseaseAndDrugs;
