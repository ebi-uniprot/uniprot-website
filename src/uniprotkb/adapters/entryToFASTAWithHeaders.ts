import { UniProtkbAPIModel, EntryType } from './uniProtkbConverter';
import { APISequenceData } from '../../tools/blast/types/apiSequenceData';

// const CHUNK_OF_TEXT_OF_SIXTY_CHARACTERS = /(.{1,60})/g;
// const CHUNK_OF_TEXT_OF_TEN_CHARACTERS = /(.{1,10})/g;

// NOTE: if we decide to do formatting here, use formatting logic from franklin
// .split(CHUNK_OF_TEXT_OF_SIXTY_CHARACTERS)
// .filter(Boolean)
// .map((line) =>
//   line.replace(CHUNK_OF_TEXT_OF_TEN_CHARACTERS, '$1 ').trim()
// )
// .join('\n');
// build a "nicely"-formatted FASTA string
// See https://www.uniprot.org/help/fasta-headers for current headers
const entryToFASTAWithHeaders = (
  entry: UniProtkbAPIModel | APISequenceData
): string => {
  let sequence = entry.sequence.value || '';
  try {
    let db;
    switch (entry.entryType) {
      case EntryType.REVIEWED:
        db = 'sp';
        break;
      case EntryType.UNREVIEWED:
        db = 'tr';
        break;
      default:
        db = '??';
    }
    let optionalProteinName = '';
    if (entry?.proteinDescription?.recommendedName?.fullName) {
      optionalProteinName = `${entry.proteinDescription.recommendedName.fullName.value} `;
    }
    let optionalOS = '';
    if (entry?.organism?.scientificName) {
      optionalOS = `OS=${entry.organism.scientificName} `;
    }
    let optionalOX = '';
    if (entry?.organism?.taxonId) {
      optionalOX = `OX=${entry.organism.taxonId} `;
    }
    let optionalGN = '';
    if (entry.genes?.[0]?.geneName?.value) {
      optionalGN = `GN=${entry.genes[0].geneName.value} `;
    }
    const pe = entry.proteinExistence[0];
    let optionalSV = '';
    if (entry?.entryAudit?.sequenceVersion) {
      optionalSV = `SV=${entry.entryAudit.sequenceVersion}`;
    }
    sequence = `>${db}|${entry.primaryAccession}|${entry.uniProtkbId} ${optionalProteinName}${optionalOS}${optionalOX}${optionalGN}PE=${pe} ${optionalSV}\n${sequence}`;
  } catch {
    /* */
  }
  return sequence;
};

export default entryToFASTAWithHeaders;
