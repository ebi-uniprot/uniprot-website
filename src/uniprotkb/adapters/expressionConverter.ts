import { CommentType } from '../types/commentTypes';
import KeywordCategory from '../types/keywordCategory';
import EntrySection from '../types/entrySection';
import { convertSection } from './sectionConverter';
import { UniProtkbAPIModel } from './uniProtkbConverter';
import { Xref } from '../../shared/types/apiModel';
import { DatabaseInfoMaps } from '../utils/database';

const keywordsCategories: KeywordCategory[] = ['Developmental stage'];

const commentsCategories: CommentType[] = [
  'TISSUE SPECIFICITY',
  'DEVELOPMENTAL STAGE',
  'INDUCTION',
];

const convertExpression = (
  data: UniProtkbAPIModel,
  databaseInfoMaps: DatabaseInfoMaps,
  uniProtKBCrossReferences?: Xref[]
) =>
  convertSection(
    data,
    databaseInfoMaps,
    commentsCategories,
    keywordsCategories,
    undefined,
    EntrySection.Expression,
    uniProtKBCrossReferences
  );

export default convertExpression;
