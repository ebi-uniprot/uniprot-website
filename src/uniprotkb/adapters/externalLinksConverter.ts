import { UIModel } from './sectionConverter';
import { UniProtkbAPIModel } from './uniProtkbConverter';
import { getXrefsForSection } from '../utils/xrefUtils';
import EntrySection from '../types/entrySection';
import { Xref } from '../../shared/types/apiModel';

const convertExternalLinks = (
  data: UniProtkbAPIModel,
  uniProtKBCrossReferences?: Xref[]
) => {
  const convertedData: UIModel = {
    commentsData: new Map(),
    keywordData: [],
    featuresData: [],
    xrefData: [],
  };
  const { comments, genes, organism } = data;

  if (comments) {
    convertedData.commentsData.set(
      'WEB RESOURCE',
      comments.filter((comment) => comment.commentType === 'WEB RESOURCE')
    );
  }
  if (uniProtKBCrossReferences) {
    convertedData.xrefData = getXrefsForSection(
      uniProtKBCrossReferences,
      EntrySection.ExternalLinks,
      genes,
      organism?.commonName
    );
  }

  return convertedData;
};
export default convertExternalLinks;
