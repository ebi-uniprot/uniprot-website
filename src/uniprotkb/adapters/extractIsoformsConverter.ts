import { AlternativeProductsComment, CommentType } from '../types/commentTypes';
import { UniProtkbAPIModel } from './uniProtkbConverter';

const extractIsoforms = (data: UniProtkbAPIModel) => {
  const alternativeProducts = data.comments?.find(
    (comment) => comment.commentType === CommentType.ALTERNATIVE_PRODUCTS
  ) as AlternativeProductsComment;
  const isoforms = alternativeProducts?.isoforms
    .map((isoform) => isoform.isoformIds.map((isoformId) => isoformId))
    .flat();
  return { isoforms: isoforms || [] };
};

export default extractIsoforms;
