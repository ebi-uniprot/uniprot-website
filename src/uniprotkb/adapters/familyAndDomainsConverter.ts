import KeywordCategory from '../types/keywordCategory';
import { FamilyAndDomainsFeatures } from '../types/featureType';
import EntrySection from '../types/entrySection';
import { CommentType } from '../types/commentTypes';
import { convertSection } from './sectionConverter';
import { UniProtkbAPIModel } from './uniProtkbConverter';
import { Xref } from '../../shared/types/apiModel';
import { UniProtKBColumn } from '../types/columnTypes';
import { DatabaseInfoMaps } from '../utils/database';

const keywordsCategories: KeywordCategory[] = ['Domain'];

export const familyAndDomainsFeaturesToColumns: Readonly<
  Record<FamilyAndDomainsFeatures, UniProtKBColumn>
> = {
  Domain: UniProtKBColumn.ftDomain,
  Region: UniProtKBColumn.ftRegion,
  Repeat: UniProtKBColumn.ftRepeat,
  Motif: UniProtKBColumn.ftMotif,
  'Compositional bias': UniProtKBColumn.ftCompbias,
  'Zinc finger': UniProtKBColumn.ftZnFing,
};

const featuresCategories = Object.keys(
  familyAndDomainsFeaturesToColumns
) as FamilyAndDomainsFeatures[];

const familyAndDomainsComments: CommentType[] = ['DOMAIN', 'SIMILARITY'];

const convertFamilyAndDomains = (
  data: UniProtkbAPIModel,
  databaseInfoMaps: DatabaseInfoMaps,
  uniProtKBCrossReferences?: Xref[]
) =>
  convertSection(
    data,
    databaseInfoMaps,
    familyAndDomainsComments,
    keywordsCategories,
    featuresCategories,
    EntrySection.FamilyAndDomains,
    uniProtKBCrossReferences
  );

export default convertFamilyAndDomains;
