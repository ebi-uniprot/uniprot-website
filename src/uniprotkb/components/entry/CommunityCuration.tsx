import {
  ExternalLink,
  HeroContainer,
  CommunityAnnotationIcon,
  ChevronDownIcon,
  Card,
  InfoList,
} from 'franklin-sites';
import cn from 'classnames';

import useDatabaseInfoMaps from '../../../shared/hooks/useDatabaseInfoMaps';

import { processUrlTemplate } from '../../../shared/utils/xrefs';
import externalUrls from '../../../shared/config/externalUrls';

import {
  CommunityAnnotation,
  Reference,
} from '../../../supporting-data/citations/adapters/citationsConverter';
import EntrySection from '../../types/entrySection';

import styles from './styles/community-curation.module.scss';

import ORCIDiDLogo from '../../../images/ORCIDiD_icon.png';
import { defaultdict } from '../../../shared/utils/utils';

const annotationGetter = (section: EntrySection) => {
  switch (section) {
    case EntrySection.Function:
      return (communityAnnotation?: CommunityAnnotation) =>
        communityAnnotation?.function;
    case EntrySection.DiseaseVariants:
    case EntrySection.PhenotypesVariants:
      return (communityAnnotation?: CommunityAnnotation) =>
        communityAnnotation?.disease;
    case EntrySection.NamesAndTaxonomy:
      return (communityAnnotation?: CommunityAnnotation) =>
        communityAnnotation?.proteinOrGene;
    default:
      return null;
  }
};

const groupByCommunityAnnotation = (
  section: EntrySection,
  communityReferences: Reference[]
) => {
  const getAnnotation = annotationGetter(section);
  if (!getAnnotation) {
    return null;
  }
  const annotationToCommunityReferences = new Map<string, Reference[]>();
  for (const communityReference of communityReferences) {
    const annotation = getAnnotation(communityReference.communityAnnotation);
    if (annotation) {
      const annotations = annotationToCommunityReferences.get(annotation) || [];
      annotations.push(communityReference);
      annotationToCommunityReferences.set(annotations);
    }
  }
  return annotationToCommunityReferences;
};

const CommunityCuration = ({
  accession,
  section,
  communityReferences,
}: {
  accession: string;
  section: EntrySection;
  communityReferences: Reference[];
}) => {
  const databaseInfoMaps = useDatabaseInfoMaps();
  const groupedCommunityReferences = groupByCommunityAnnotation(
    section,
    communityReferences
  );
  if (groupedCommunityReferences?.size) {
    return (
      <details className={styles['community-annotation-details']}>
        <summary className={styles.header}>
          <span
            className={cn('button', 'tertiary', styles['community-button'])}
          >
            <CommunityAnnotationIcon width="1ch" />
            {`Community curation (${communityReferences.length}) `}
            <ChevronDownIcon width="1ch" className={styles['chevron-icon']} />
          </span>
          <hr className={styles.separator} />
        </summary>
        <HeroContainer className={styles.content}>
          {communityReferences.map(
            ({ source, communityAnnotation, citationId }) => {
              const contributorElement = (
                <div
                  className={cn(
                    styles['contributor-details'],
                    section === EntrySection.NamesAndTaxonomy
                      ? styles['names-section']
                      : ''
                  )}
                >
                  {citationId && (
                    <span>
                      Source:&nbsp;&nbsp;
                      <ExternalLink url={externalUrls.PubMed(citationId)}>
                        PMID - {citationId}
                      </ExternalLink>
                    </span>
                  )}
                  {communityAnnotation?.submissionDate && (
                    <span>
                      Submission date:&nbsp;&nbsp;
                      {communityAnnotation?.submissionDate}
                    </span>
                  )}
                  {source && (
                    <span>
                      Contributor:&nbsp;&nbsp;
                      {source.id && source.id !== 'Anonymous' ? (
                        <ExternalLink
                          url={processUrlTemplate(
                            databaseInfoMaps?.databaseToDatabaseInfo[
                              source.name
                            ]?.uriLink,
                            { id: source.id }
                          )}
                        >
                          <img
                            src={ORCIDiDLogo}
                            alt=""
                            width="15"
                            height="15"
                          />
                          {source.id}
                        </ExternalLink>
                      ) : (
                        source.id
                      )}
                    </span>
                  )}
                  <ExternalLink
                    url={externalUrls.CommunityCurationGet(accession)}
                  >
                    View submission
                  </ExternalLink>
                </div>
              );

              return (
                <Card key={citationId} className={styles['reference-card']}>
                  {(section === EntrySection.Function ||
                    section === EntrySection.PhenotypesVariants ||
                    section === EntrySection.DiseaseVariants) && (
                    <>
                      <p>
                        {section === EntrySection.Function &&
                          communityAnnotation?.function}
                        {(section === EntrySection.DiseaseVariants ||
                          section === EntrySection.PhenotypesVariants) &&
                          communityAnnotation?.disease}
                      </p>
                      {contributorElement}
                    </>
                  )}
                  {section === EntrySection.NamesAndTaxonomy && (
                    <InfoList
                      infoData={[
                        {
                          title: 'Community suggested names',
                          content: (
                            <>
                              {communityAnnotation?.proteinOrGene}
                              {contributorElement}
                            </>
                          ),
                        },
                      ]}
                    />
                  )}
                </Card>
              );
            }
          )}
        </HeroContainer>
      </details>
    );
  }
  return null;
};

export default CommunityCuration;
