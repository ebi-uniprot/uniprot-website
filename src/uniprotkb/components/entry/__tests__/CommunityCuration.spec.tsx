import { screen } from '@testing-library/react';

import CommunityCuration from '../CommunityCuration';

import customRender from '../../../../shared/__test-helpers__/customRender';

import { Reference } from '../../../../supporting-data/citations/adapters/citationsConverter';
import EntrySection from '../../../types/entrySection';

const mock: Reference[] = [
  {
    source: {
      name: 'ORCID',
      id: '0000-0001-6057-9374',
    },
    citationId: '36301857',
    sourceCategories: [
      'Function',
      'Expression',
      'Interaction',
      'Subcellular Location',
    ],
    communityAnnotation: {
      proteinOrGene: 'Epicuticlin',
      function:
        'Disordered proteins organized in tandem repeats with molecular recognition features and tyrosine motifs (Pfam02756 and Pfam02757). The epicuticlins form the insoluble outermost layer of the cuticle and can interact with cuticular collagens.',
      comment:
        'The cDNA AJ408886 is one of several sequences of nematodes, which is now identified as forming the epicuticlin structure of nematodes.',
    },
  },
];

describe('Community annotatation', () => {
  it('should render the community annotation content', async () => {
    const { asFragment } = customRender(
      <CommunityCuration
        accession="Q95PB5"
        section={EntrySection.NamesAndTaxonomy}
        communityReferences={mock}
      />
    );
    expect(await screen.findByText(/Epicuticlin/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
