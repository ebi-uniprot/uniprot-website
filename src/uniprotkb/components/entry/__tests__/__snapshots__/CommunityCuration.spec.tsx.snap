// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Community annotatation should render the community annotation content 1`] = `
<DocumentFragment>
  <details
    class="community-annotation-details"
  >
    <summary
      class="header"
    >
      <span
        class="button tertiary community-button"
      >
        <test-file-stub
          width="1ch"
        />
        Community curation (1) 
        <test-file-stub
          classname="chevron-icon"
          width="1ch"
        />
      </span>
      <hr
        class="separator"
      />
    </summary>
    <section
      class="hero-container content hero-container--side-padding"
    >
      <section
        class="reference-card card"
      >
        <div
          class="card__container"
        >
          <div
            class="card__content"
          >
            <ul
              class="info-list"
            >
              <li>
                <div
                  class="decorated-list-item"
                >
                  <div
                    class="decorated-list-item__title tiny"
                  >
                    Community suggested names
                  </div>
                  <div
                    class="decorated-list-item__content"
                  >
                    Epicuticlin
                    <div
                      class="contributor-details names-section"
                    >
                      <span>
                        Source:  
                        <a
                          class="external-link"
                          href="https://pubmed.ncbi.nlm.nih.gov/36301857"
                          rel="noopener"
                          target="_blank"
                        >
                          PMID - 36301857
                          <test-file-stub
                            data-testid="external-link-icon"
                            width="12.5"
                          />
                        </a>
                      </span>
                      <span>
                        Contributor:  
                        <a
                          class="external-link"
                          href="https://orcid.org/0000-0001-6057-9374"
                          rel="noopener"
                          target="_blank"
                        >
                          <img
                            alt=""
                            height="15"
                            src="test-file-stub"
                            width="15"
                          />
                          0000-0001-6057-9374
                          <test-file-stub
                            data-testid="external-link-icon"
                            width="12.5"
                          />
                        </a>
                      </span>
                      <a
                        class="external-link"
                        href="https://community.uniprot.org/cgi-bin/bbsub_query?accession=Q95PB5"
                        rel="noopener"
                        target="_blank"
                      >
                        View submission
                        <test-file-stub
                          data-testid="external-link-icon"
                          width="12.5"
                        />
                      </a>
                    </div>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </section>
    </section>
  </details>
</DocumentFragment>
`;
